e2e/vaultcompat/vaultcompat_test.go

```diff
@@ -0,0 +1,260 @@
+	log.Write("auth/token/roles/nomad-cluster", role)
+
+	token := vc.Auth().Token()
+	secret, err := token.Create(&vaultapi.TokenCreateRequest{
+		Policies: []string{"nomad-server"},
+		Period:   "72h",
+		NoParent: true,
+	})
+	must.NoError(t, err, must.Sprint("failed to create vault token"))
+	must.NotNil(t, secret)
+	must.NotNil(t, secret.Auth)
+}
+
+func startNomad(t *testing.T, vc *vaultapi.Client) (func(), *nomadapi.Client) {
+	ts := testutil.NewTestServer(t, func(c *testutil.TestServerConfig) {
+		c.Vault = &testutil.VaultConfig{
+			Enabled:              true,
+			Address:              vc.Address(),
+			Token:                vc.Token(),
+			Role:                 "nomad-cluster",
+			AllowUnauthenticated: true,
+		}
+		c.DevMode = true
+		c.Client = &testutil.ClientConfig{
+			Enabled: true,
+		}
+		c.LogLevel = "off"
+	})
+	nc, err := nomadapi.NewClient(&nomadapi.Config{
+		Address: "http://" + ts.HTTPAddr,
+	})
+	must.NoError(t, err, must.Sprint("unable to create nomad api client"))
+	return ts.Stop, nc
+}
+
+func downloadVaultBuild(t *testing.T, b build) {
+	path := filepath.Join(os.TempDir(), binDir, b.Version)
+	must.NoError(t, os.MkdirAll(path, 0755))
+
+	if _, err := os.Stat(filepath.Join(path, "vault")); !os.IsNotExist(err) {
+		t.Log("download: already have vault at", path)
+		return
+	}
+
+	t.Log("download: installing vault at", path)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, "hc-install", "install", "-version", b.Version, "-path", path, "vault")
+	bs, err := cmd.CombinedOutput()
+	must.NoError(t, err, must.Sprintf("failed to download vault %s: %s", b.Version, string(bs)))
+}
+
+func getMinimumVersion(t *testing.T) *version.Version {
+	v, err := version.NewVersion("1.1.0")
+	must.NoError(t, err)
+	return v
+}
+
+type build struct {
+	Version string `json:"version"`
+	OS      string `json:"os"`
+	Arch    string `json:"arch"`
+	URL     string `json:"url"`
+}
+
+func (b build) String() string { return b.Version }
+
+func (b build) compare(o build) int {
+	B := version.Must(version.NewVersion(b.Version))
+	O := version.Must(version.NewVersion(o.Version))
+	return B.Compare(O)
+}
+
+type vaultJSON struct {
+	Versions map[string]struct {
+		Builds []build `json:"builds"`
+	}
+}
+
+func usable(v, minimum *version.Version) bool {
+	switch {
+	case v.Prerelease() != "":
+		return false
+	case v.Metadata() != "":
+		return false
+	case v.LessThan(minimum):
+		return false
+	default:
+		return true
+	}
+}
+
+func keep(b build) bool {
+	switch {
+	case b.OS != runtime.GOOS:
+		return false
+	case b.Arch != runtime.GOARCH:
+		return false
+	default:
+		return true
+	}
+}
+
+type tracker map[int]*set.TreeSet[build, set.Compare[build]]
+
+func (t tracker) add(v *version.Version, b build) {
+	y := v.Segments()[1]
+	if _, exists := t[y]; !exists {
+		cmp := func(g, h build) int { return g.compare(h) }
+		t[y] = set.NewTreeSet[build, set.Compare[build]](cmp)
```

Comment:
I had to churn some extra brain cells to parse this, compared to everything else in here being very readable, but the end result of `track.add()` and then `tree.Max()` down below (if I'm reading this all right) is quite nice. :+1: